### PR TITLE
Database Schema: Base-64 Encode Data Names and Home Domains

### PR DESCRIPTION
**Pull request by jonjove**
Database schema upgrade to base-64 encode data names and home domains.

### CODE REVIEW

File: src/ledger/LedgerTxnDataSQL.cpp [hunk 1]
```diff
@@ -127,4 +127,71 @@ LedgerTxnRoot::Impl::dropData()
...
+    mEntryCache.clear();
+    mBestOffersCache.clear();
+
+    mDatabase.getSession()
+        << "ALTER TABLE accountdata RENAME TO accountdataunencoded;";
+    dropData();
+
+    size_t const LIMIT = 4096;
+    size_t offset = 0;
+    auto dataToEncode = loadDataToEncode(mDatabase, LIMIT, offset);
```
Comment (MonsieurNicolas) [inline src/ledger/LedgerTxnDataSQL.cpp]: I would ditch the LIMIT offset (and loop) given that we have order of 100k data entries total

File: src/database/Database.cpp [hunk 2]
```diff
@@ -51,7 +51,7 @@ using namespace std;
 
 bool Database::gDriversRegistered = false;
 
-static unsigned long const SCHEMA_VERSION = 8;
+static unsigned long const SCHEMA_VERSION = 9;
```
Comment (MonsieurNicolas) [inline src/database/Database.cpp]: no need to bump the version: I imagine that we're going to merge this into the upcoming release

File: src/database/Database.cpp [hunk 3]
```diff
@@ -129,6 +129,11 @@ Database::applySchemaUpgrade(unsigned long vers)
         mSession << "DROP TABLE IF EXISTS signers";
         break;
 
+    case 9:
```
Comment (MonsieurNicolas) [inline src/database/Database.cpp]: this commit should be after the one that introduces `encodeHomeDomainsBase64`

File: src/ledger/LedgerTxnAccountSQL.cpp [hunk 4]
```diff
@@ -330,4 +330,70 @@ LedgerTxnRoot::Impl::dropAccounts()
         << "CREATE INDEX accountbalances ON accounts (balance) WHERE "
            "balance >= 1000000000";
 }
+
+static std::unordered_map<std::string, std::string>
+loadHomeDomainsToEncode(Database& db, size_t limit, size_t offset)
+{
+    std::string accountID, homeDomain;
+
+    std::string sql = "SELECT accountid, homedomain FROM accounts "
```
Comment (MonsieurNicolas) [inline src/ledger/LedgerTxnAccountSQL.cpp]: you can make this a lot faster by adding `WHERE length(homedomain) > 0`

File: src/ledger/LedgerTxnAccountSQL.cpp [hunk 5]
```diff
@@ -330,4 +330,70 @@ LedgerTxnRoot::Impl::dropAccounts()
...
+void
+LedgerTxnRoot::Impl::encodeHomeDomainsBase64()
+{
+    throwIfChild();
+    mEntryCache.clear();
+    mBestOffersCache.clear();
+
+    size_t const LIMIT = 4096;
+    size_t offset = 0;
+    auto homeDomainsToEncode =
```
Comment (MonsieurNicolas) [inline src/ledger/LedgerTxnAccountSQL.cpp]: same thing, you can safely ditch the LIMIT/offset and related logic, with the filtering we're talking less than 100k accounts which is safe to keep in memory

File: src/ledger/LedgerTxnAccountSQL.cpp [hunk 6]
```diff
@@ -330,4 +330,70 @@ LedgerTxnRoot::Impl::dropAccounts()
...
+
+    std::string sql =
+        "UPDATE accounts SET homedomain = :v1 WHERE accountid = :v2";
+
+    auto prep = db.getPreparedStatement(sql);
+    auto& st = prep.statement();
+    st.exchange(soci::use(encodedHomeDomain));
+    st.exchange(soci::use(accountID));
+    st.define_and_bind();
+    st.execute(true);
```
Comment (MonsieurNicolas) [inline src/ledger/LedgerTxnAccountSQL.cpp]: should we check that nb effected row is 1?

File: src/ledger/LedgerTxnAccountSQL.cpp [hunk 7]
```diff
@@ -330,4 +330,70 @@ LedgerTxnRoot::Impl::dropAccounts()
...
+    auto prep = db.getPreparedStatement(sql);
+    auto& st = prep.statement();
+    st.exchange(soci::into(accountID));
+    st.exchange(soci::into(homeDomain));
+    st.exchange(soci::use(limit));
+    st.exchange(soci::use(offset));
+    st.define_and_bind();
+    st.execute(true);
+
+    std::unordered_map<std::string, std::string> res;
```
Comment (MonsieurNicolas) [inline src/ledger/LedgerTxnAccountSQL.cpp]: nit: you don't really need a k:v type of structure for this (and return type of the function), a `std::vector<std::pair<>>` is enough

File: src/ledger/LedgerTxnDataSQL.cpp (hunk 1 above)
Comment (MonsieurNicolas) [reply src/ledger/LedgerTxnDataSQL.cpp]: technically if we do that we don't even need to keep the data around in `accountdataunencoded` (ie: we load all data, recreate the table, write the data), so much simpler logic ; one of the side effects of having 2 tables might be unnecessary overhead that will have to be vacuumed.

Comment (graydon) [reply src/ledger/LedgerTxnDataSQL.cpp]: Yeah, I agree there's unlikely to be a problem doing it all in memory.

File: src/ledger/LedgerTxnImpl.h [hunk 8]
```diff
@@ -421,5 +421,17 @@ class LedgerTxnRoot::Impl
     // - the prepared statement cache may be, but is not guaranteed to be,
     //   modified
     void writeSignersTableIntoAccountsTable();
+
+    // encodedDataNamesBase64 has the basic exception safety guarantee. If it
+    // throws an exception, then
+    // - the prepared statement cache may be, but is not guaranteed to be,
+    //   modified
+    void encodeDataNamesBase64();
```
Comment (vogel) [inline src/ledger/LedgerTxnImpl.h]: Do we need those in that class? Woudn't static methods/free function be better here? Those are one-time calls (even one-time per instance, not per launch), so why pollute that interface?

Comment (vogel) [reply src/ledger/LedgerTxnImpl.h]: OK, but I don't agree with that solution as I would like to see a layer below LedgerTxnRoot to handle the queries.

For example dropOffers does a lot of unnecessary things, like "throwIfChild" (it is only called from Database::initialize!). I think we could make it as safe as it is now with a different solution (but this is very low priority).

And if you really want all SQL statements in LedgerTxnRoot, why not move there also all of `ALTER TABLE` from liabilities and signers changes?

Comment (MonsieurNicolas) [reply src/ledger/LedgerTxnImpl.h]: yeah I would like to see some separation too (and/or limit public methods) but that is outside of the scope of this PR and seems like lower priority. Open an issue to discuss that if you feel like we should push for it.

**Comment by MonsieurNicolas** [comment]
r+ 9361012241017781fe537eff627800e9419ad39c

**Comment by MonsieurNicolas** [comment]
I opened a couple issues related to the bug that you caught in the account signers change that you added to this PR

**Comment by MonsieurNicolas** [comment]
this was reverted by https://github.com/stellar/stellar-core/pull/1933

**Pull request merged**